[benchmark] Add ReplaceSubrange benchmark

keitaito opened the pull request:
## 🔨 Changes

Add ReplaceSubrange benchmark. It is still work in progress, but it would be great if I could get feedback to make sure I'm on the right direction 🙇‍♂️

Resolves a part of [SR-8905 Gaps in String benchmarking](https://bugs.swift.org/browse/SR-8905).

Shout-out to Rob and Michael for mentoring at try! Swift San Jose!

### To-do list
- [x] Arguments of types String
- [x] Substring
- [x] Array<Character>
- [x] Repeated<Character>

keitaito committed:
Add ReplaceSubrange benchmark

brentdax commented:
@milseman I'm thinking this is in your wheelhouse.

keitaito committed:
Update ReplaceSubrange benchmark by adding Substring case

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please benchmark

brentdax commented:
@keitaito This is a great start, but can you address the issues listed at the bottom of the benchmark comment, please?

palimondo commented:
@keitaito I suggest naming these benchmarks per [naming convention](http://bit.ly/BenchmarkNaming) as
* `String.replaceSubrange.SmallLiteral`
* `String.replaceSubrange.LargeLiteral`
* `String.replaceSubrange.LargeManaged`

and the file  as `StringReplaceSubrange.swift`.

palimondo commented:
You need to half the loop multiplier. Mind the 80 character line limit. I would also just create a single testing function and pass the `string` as parameter. Since it'll be marked as `@inline(never)`, you don't need to use `getString` to prevent the optimizations. Then you'd use inline closure to define `runFunction`s. You can initialize the large managed string in setup also with inline closure and `blackHole` function:

```swift
let t: [BenchmarkCategory] = [.validation, .api, .String]

public let StringReplaceSubrange = [
  BenchmarkInfo(name: "String.replaceSubrange.SmallLiteral",
    runFunction: { replaceSubrange($0, "coffee" }, tags: t),
  BenchmarkInfo(name: "String.replaceSubrange.LargeLiteral",
    runFunction: { replaceSubrange($0, "coffee\u{301}coffeecoffeecoffee" },
    tags: t),
  BenchmarkInfo(name: "ReplaceSubrangeWithLargeManagedString",
    runFunction: { replaceSubrange($0, largeManaged) }, tags: t,
    setUpFunction: { blackHole(largeManaged) }),
]

@inline(never)
public func replaceSubrange(_ N: Int, _ string: String) {
  var copy = string
  let range = string.startIndex..<string.index(after: string.startIndex)
  for _ in 0 ..< 5_000*N {
    copy.replaceSubrange(range, with: "t")
  }
}
````

@milseman Does it make sense to test range replacement with a single character that doesn't change the size? How about something like this?

````swift
@inline(never)
public func replaceSubrange(_ N: Int, _ string: String) {
  var s = string
  for _ in 0 ..< 2_500*N {
    let ff =
      s.index(s.startIndex, offsetBy:2)..<s.index(s.startIndex, offsetBy:4)
    s.replaceSubrange(ff, with: "☕️")
    let cup =
      s.index(s.startIndex, offsetBy:2)..<s.index(s.startIndex, offsetBy:3)
    copy.replaceSubrange(cup, with: "ff")
  }
  CheckResults(s == string)
}
````

Also, since `replaceSubrange` is mutable operation, how much sense does it make to vary these across literals and managed strings? After the initial modification, wouldn't this differ only in `StringGuts`(`_SmallString` or not)?

palimondo commented:
@milseman Could you also please verify my reasoning above regarding the `getString` and shared run function? I'm starting to doubt myself looking at the reported runtimes…

keitaito commented:
@brentdax Thanks Brent for running tests 😄

@palimondo Hi Pavol, thanks for the great feedback! I will start working on easy fixes like re-naming the file and benchmarks, and fixing line length 🔨

keitaito committed:
Rename ReplaceSubrange to StringReplaceSubrange

keitaito committed:
Rename benchmarks based on the benchmark naming convention

Reference: https://github.com/apple/swift/blob/master/benchmark/Naming.md

milseman commented:
@palimondo 

> I would also just create a single testing function and pass the string as parameter. Since it'll be marked as @inline(never), you don't need to use getString to prevent the optimizations.

I don't think `@inline(never)` is sufficient here, but I agree with having a single function that runs over different workloads. You could say:

```swift
@inline(never)
public func replaceSubrange(_ N: Int, _ string: String) {
  var copy = getString(string)
  let range = string.startIndex..<string.index(after: string.startIndex)
  for _ in 0 ..< 5_000*N {
    copy.replaceSubrange(range, with: "t")
  }
}
```

So there's no measurable overhead in the getString call, but it serves as an optimization barrier, should the optimizer ever do more inter-procedural optimizations that are not contingent on prior inlining.

> Also, would it make sense to do a variant that would cross the boundary between the small and large representations? Or vary the benchmarks based on the position of the range, i.e. replacing from start, in the middle or just at the end, to stress the reallocations?

I think *a* benchmark that grows a small string to a large one could be interesting. We don't often go the other way, and we don't shrink the string under normal operations (but `keepingCapacity: false`, making a brand new one, etc., will form the small string). If a string is already allocated with excess capacity, it's likely the user requested that capacity or are otherwise going to be filling it soon, so there's more harm in deallocating it right away only to re-allocate.

I don't think it matters what that operation is, and `append` is more straightforward for a micro-benchmark than `replaceSubrange`.

milseman reviewed (commented):
This looks good if you can incorporate @palimondo 's feedback.

After thinking about this more, the `largeLiteral` is likely redundant with `largeManaged`, because after the very first `replaceSubrange()` call, it will be managed (due to mutation) for the subsequent `10_000` operations. The large vs small split is nice, though.

keitaito committed:
Reuse replaceSubrange(_:_:with:) function for benchmarks

keitaito committed:
Add benchmarks for String.replaceSubrange(_:with:) with Array<Character> arguments

keitaito committed:
Add benchmarks for String.replaceSubrange(_:with:) with Repeated<Character>

keitaito committed:
Remove benchmarks with largeLiteral

Per Michael's feedback (https://github.com/apple/swift/pull/25310#pullrequestreview-248466570), largeLiteral is likely redundant with largeManaged.

keitaito commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+
+private var largeManagedString: String = {
+    return getString("coffee\u{301}coffeecoffeecoffeecoffee")
+}()
```
`\u{301}` was suggested to be added to the test string when I paired with Michael. @milseman would you mind reminding me the reason for this? I don't remember it now 😅

palimondo replied:
I'll guess it was in order for the string to be in particular [normalization form](https://unicode.org/reports/tr15/). @milseman Do you want to vary the benchmarks also for different normalization forms? [SR-8905](https://bugs.swift.org/browse/SR-8905) doesn't mention that…

keitaito replied:
I was wondering if the result of testing with the string "coffeécoffeecoffeecoffeecoffee" would be different from the one with "coffeecoffeecoffeecoffeecoffee" 🤔 If there is distinct difference, maybe we could add two benchmarks for the one with the acute accent character and the other one without it?

milseman replied:
Grapheme segmentation is relevant to this benchmark, but not normalization (since there's no comparison). The difference is that the precomposed representation is a single scalar per grapheme cluster, while the decomposed (multi-scalar) form is not. The single-scalar one will hit our grapheme breaking fast-paths while the multi-scalar one will call out to ICU. Alternatively, you could use other kinds of multi-scalar graphemes clusters, such as complex emoji. I just mentioned `"\u{301}"` because you can just stick it after an `"e"` to get the same effect.

keitaito commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.RepeatedChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getRepeatedCharacter(repeatedCharacter)) },
+    tags: tags
```
The benchmark name "Str.replaceSubrange.SmallLiteral.RepeatedChar" is longer than 40 characters, but I couldn't think a better name fitting 40. Maybe it can be like "Str.replaceSubrange.LargeManagedRepChar", but I was concerned "RepChar" is a little bit hard to understand that it means `Repeated<Character>`. @palimondo What do you think about this naming? Please let me know if you have any suggestions on it 🙂

palimondo replied:
Looking at what @milseman writes in [SR-8905](https://bugs.swift.org/browse/SR-8905):

> replaceSubrange<C: Collection>(_:C)
> * Arguments of types `String`, `Substring`, `Array<Character>`, `Repeated<Character>`, etc

I'd say the [naming convention](http://bit.ly/BenchmarkNaming) calls for base name of `String.replaceSubrange` which varies across the argument type (`String`, `Substring`, `ArrChar`, `RepChar`)  for the general case of large strings. Then we'll denote the special optimization for small strings with a simple `.Small` suffix and we'll get these benchmarks:

* `String.replaceSubrange.String`
* `String.replaceSubrange.Substring`
* `String.replaceSubrange.ArrChar`
* `String.replaceSubrange.RepChar`
* `String.replaceSubrange.String.Small`
* `String.replaceSubrange.Substring.Small`
* `String.replaceSubrange.ArrChar.Small`
* `String.replaceSubrange.RepChar.Small`

The longest one is `String.replaceSubrange.Substring.Small` at 39 characters, just under the 40 chars limit.

keitaito replied:
That's an awesome naming idea. I will use them. Thanks for your suggestion!

keitaito commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+
+private func setupLargeManagedString() {
+    _ = largeManagedString
+}
```
Should `largeManagedString` be called from `setupFunction` closure  before it is used for benchmarks?

palimondo replied:
Given that in the `replaceSubrange(::with:)` you already do the `var copy = getString(string)`, this whole dance here is unnecessary. You should declare this as simple `let largeString = "coffee\u{301}coffeecoffeecoffeecoffee"` and drop all the `setUpFunction`s.

keitaito replied:
That makes sense. Thanks for explaining!

keitaito commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+  let range = string.startIndex..<string.index(after: string.startIndex)
+  for _ in 0 ..< 5_000 * N {
+    copy.replaceSubrange(range, with: replacingString)
+  }
```
What is the criteria for choosing this multiplying number like 5000? Does this depend on the benchmark time?

palimondo replied:
Correct. We are just trying to size the workload to run in 20–1000 μs, so that it is in a measurement sweet spot for our system.

keitaito replied:
Thanks for explaining!

keitaito commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.String",
+    runFunction: { replaceSubrange($0, "coffee", with: "t") },
+    tags: tags
```
Would you mind explaining what is the difference between the small literal string vs the large managed string? Is this something related to [this small string optimization](https://gist.github.com/milseman/bb39ef7f170641ae52c13600a512782f#small-string-optimization)? If the string fits 15 ASCII characters length, it won't be allocated in the heap memory?

palimondo replied:
Correct. See: [`_SmallString`](https://github.com/apple/swift/blob/master/stdlib/public/core/SmallString.swift) and  [`_StringGuts`](https://github.com/apple/swift/blob/master/stdlib/public/core/StringGuts.swift) for implementation details if you're interested.

keitaito replied:
Thanks for the link! I will take a look at them 🔍

milseman replied:
Small form accommodates 15 UTF-8 code units in length (not just ASCII)

keitaito replied:
Thank you for clarifying the length of the small string, Michael 😄

keitaito reviewed (commented):
Sorry for taking a while to update this PR, but now this PR adds `String.replaceSubrange(_:with:)` benchmarks with arguments of types String, Substring, Array<Character>, and Repeated<Character>. Would you mind taking a look at this PR again when you have time?

Also, while I was working on this PR, I've got a few questions related to Swift.String. Would you mind giving me some explanations on them if you have time? 🙇‍♂️

palimondo commented:
@swift-ci please benchmark

palimondo commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.Substr",
+    runFunction: { replaceSubrange($0, "coffee", with: getSubstring("t")) },
```
We don't need to put an optimization barrier (by calling the `getSubstring` from `TestUtils`) here.
Shorter way to get a `Substring` is to get a full subrange like this: `"t"[...]`.

For an implementation symmetry, I'd also extract the "coffee" into `smallString` constant, so that our benchmark definitions would vary only like this:
* `runFunction: { replaceSubrange($0, largeString, with: "t") }`
* `runFunction: { replaceSubrange($0, smallString, with: "t") }`
* `runFunction: { replaceSubrange($0, largeString, with: "t"[...]) }`
* `runFunction: { replaceSubrange($0, smallString, with: "t"[...]) }`

palimondo commented on benchmark/utils/TestsUtils.swift:
```diff
@@ -322,3 +322,11 @@ public func getString(_ s: String) -> String { return s }
 // The same for Substring.
 @inline(never)
 public func getSubstring(_ s: Substring) -> Substring { return s }
+
```
I don't think we need to define any new optimization barrier functions...

palimondo reviewed (changes requested):
I think one generic test method should be able to over all of our test cases. See inline comments.

palimondo commented on benchmark/single-source/StringReplaceSubrange.swift:
```diff
@@ -0,0 +1,116 @@
+}
+
+@inline(never)
+private func replaceSubrange(_ N: Int, _ string: String, with replacingSubstring: Substring) {
```
If I understand @milseman's intention from [SR-8905](https://bugs.swift.org/browse/SR-8905) correctly, we are designing benchmark for the generic `replaceSubstring` method that varies across the `String`, `Substring`, `Array<Character>` and `Repeat<Character>` specializations.

Therefore we should be able to define single shared generic test function and vary the parameter in `runFunction` closure in `BenchmarkInfo` declarations.

For an example of such benchmarks, see [`append`](https://github.com/apple/swift/blob/1cbc0b376fdcd6af16d201b35fa568fdf29bb305/benchmark/single-source/DataBenchmarks.swift#L453-L460) variants in [`DataBenchmarks`](https://github.com/apple/swift/blob/1cbc0b376fdcd6af16d201b35fa568fdf29bb305/benchmark/single-source/DataBenchmarks.swift#L233-L238)

@milseman Any thoughts on keeping or dropping the `@inline(never)` annotation?

milseman replied:
You could refactor this to a generic `@inline(__always)` implementation function that takes a scale factor (because the String version should be much faster than the Repeat version) and does the loop and `replaceSubrange` calls. You'd likely want to make a `@inline(never)` top-level function for each individual benchmark. It doesn't save you a whole lot.

palimondo reviewed (changes requested):
This review comment got somehow lost… Let me try again.

keitaito commented:
@palimondo Thank you so much for another review, Pavol! I will address your comments and fix warnings and errors from the swift-ci benchmark check report 🔨

keitaito commented:
Several build failure errors are suddenly thrown when running `ninja swift-benchmark-macosx-x86_64` 😢 Have you seen these errors? Do you have any ideas how to fix them and re-build swift-benchmark?

```
$ ninja swift-benchmark-macosx-x86_64

swift-source/swift/stdlib/public/stubs/MathStubs.cpp:21:10: fatal error: 'climits' file not found
swift-source/swift/include/swift/Runtime/Debug.h:20:10: fatal error: 'cstdarg' file not found
swift-source/swift/stdlib/public/stubs/CommandLine.cpp:17:10: fatal error: 'vector' file not found
swift-source/swift/stdlib/public/stubs/LibcShims.cpp:30:10: fatal error: 'type_traits' file not found
swift-source/swift/include/swift/Runtime/Debug.h:20:10: fatal error: 'cstdarg' file not found
swift-source/swift/stdlib/public/stubs/ThreadLocalStorage.cpp:13:10: fatal error: 'cstring' file not found
swift-source/swift/stdlib/public/stubs/Stubs.cpp:37:10: fatal error: 'climits' file not found
swift-source/swift/stdlib/public/stubs/../SwiftShims/RefCount.h:30:10: fatal error: 'type_traits' file not found
swift-source/swift/include/swift/Basic/Lazy.h:16:10: fatal error: 'memory' file not found

swift-source/swift/stdlib/public/runtime/ImageInspection.h:25:10: fatal error: 'cstdint' file not found
swift-source/swift/include/swift/ABI/Metadata.h:20:10: fatal error: 'atomic' file not found
swift-source/llvm/include/llvm/Support/Compiler.h:20:10: fatal error: 'new' file not found
swift-source/swift/include/swift/Runtime/HeapObject.h:20:10: fatal error: 'cstddef' file not found
swift-source/swift/include/swift/Basic/Range.h:38:10: fatal error: 'algorithm' file not found
swift-source/swift/include/swift/Demangling/Demangle.h:22:10: fatal error: 'memory' file not found
swift-source/swift/include/swift/Basic/Lazy.h:16:10: fatal error: 'memory' file not found
swift-source/swift/include/swift/Basic/Lazy.h:16:10: fatal error: 'memory' file not found
swift-source/swift/include/swift/ABI/Metadata.h:20:10: fatal error: 'atomic' file not found
swift-source/swift/include/swift/ABI/Metadata.h:20:10: fatal error: 'atomic' file not found
```

A lot of error messages are displayed on Terminal. Here is an example of a build failure error. 
```
[1/743] Building CXX object stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Enum.cpp.o
FAILED: stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Enum.cpp.o 
/[my-own-path]/swift-source/build/Ninja-RelWithDebInfoAssert/llvm-macosx-x86_64/./bin/clang++  -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -D_DEBUG -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Istdlib/public/runtime -I/[my-own-path]/swift-source/swift/stdlib/public/runtime -Iinclude -I/[my-own-path]/swift-source/swift/include -I/[my-own-path]/swift-source/llvm/include -I/[my-own-path]/swift-source/build/Ninja-RelWithDebInfoAssert/llvm-macosx-x86_64/include -I/[my-own-path]/swift-source/llvm/tools/clang/include -I/[my-own-path]/swift-source/build/Ninja-RelWithDebInfoAssert/llvm-macosx-x86_64/tools/clang/include -I/[my-own-path]/swift-source/cmark/src -I/[my-own-path]/swift-source/build/Ninja-RelWithDebInfoAssert/cmark-macosx-x86_64/src -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wstring-conversion -fdiagnostics-color -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -fno-sanitize=all -DLLVM_DISABLE_ABI_BREAKING_CHECKS_ENFORCING=1 -O2  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk   -UNDEBUG  -fno-exceptions -fno-rtti -Wall -Wglobal-constructors -Wexit-time-destructors -fvisibility=hidden -DSWIFT_RUNTIME_CLOBBER_FREED_OBJECTS=1 -DswiftCore_EXPORTS -I/[my-own-path]/swift-source/swift/include -DSWIFT_TARGET_LIBRARY_NAME=swiftRuntime -target x86_64-apple-macosx10.9 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -arch x86_64 -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/../../../Developer/Library/Frameworks -mmacosx-version-min=10.9 -O2 -g -UNDEBUG -DSWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS -MD -MT stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Enum.cpp.o -MF stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Enum.cpp.o.d -o stdlib/public/runtime/CMakeFiles/swiftRuntime-macosx-x86_64.dir/Enum.cpp.o -c /[my-own-path]/swift-source/swift/stdlib/public/runtime/Enum.cpp
In file included from /[my-own-path]/swift-source/swift/stdlib/public/runtime/Enum.cpp:17:
In file included from /[my-own-path]/swift-source/llvm/include/llvm/Support/ErrorHandling.h:18:
/[my_path]swift-source/llvm/include/llvm/Support/Compiler.h:20:10: fatal error: 'new' file not found
#include <new>
         ^~~~~
1 error generated.
```

I'm not sure if it's related, but these errors started being thrown after Xcode 11 beta 3 was installed on my machine, FWIW.

palimondo commented:
I'd guess you need to do a clean build after Xcode update.

keitaito committed:
Update the PR by addressing feedback

- Re-name benchmark names
- Remove setupLargeManagedString()
- Remove unnecessary optimization barrier functions
- Update benchmark argument string by using smallString and largeString
- Remove unnecessary optimization barrier function calls
- Update replaceSubrange(_:_:with) to be generic function

keitaito commented:
@palimondo Thanks for the reply on the build issue, Pavol! I ended up re-building everything from scratch 😅

keitaito commented:
The benchmark with `Repeated<Character>` takes much longer time than others. Do you have any idea why this type is slower than others? Is there any way to improve this benchmark? The benchmark result on my machine is the following:

```
// When the for-in loop runs with `0 ..< 5_000 * N`
$ ./Benchmark_O 772 773 774 775 776 777 778 779
#,TEST,SAMPLES,MIN(μs),MAX(μs),MEAN(μs),SD(μs),MEDIAN(μs)
772,String.replaceSubrange.ArrChar,1,2049,2049,2049,0,2049
773,String.replaceSubrange.ArrChar.Small,1,1240,1240,1240,0,1240
774,String.replaceSubrange.RepChar,1,31690,31690,31690,0,31690
775,String.replaceSubrange.RepChar.Small,1,4670,4670,4670,0,4670
776,String.replaceSubrange.String,1,1105,1105,1105,0,1105
777,String.replaceSubrange.String.Small,1,1233,1233,1233,0,1233
778,String.replaceSubrange.Substring,1,2771,2771,2771,0,2771
779,String.replaceSubrange.Substring.Small,1,1477,1477,1477,0,1477

Total performance tests executed: 8
```

```
// When the for-in loop runs with `0 ..< 250 * N`
$ ./Benchmark_O 772 773 774 775 776 777 778 779
#,TEST,SAMPLES,MIN(μs),MAX(μs),MEAN(μs),SD(μs),MEDIAN(μs)
772,String.replaceSubrange.ArrChar,1,92,92,92,0,92
773,String.replaceSubrange.ArrChar.Small,1,54,54,54,0,54
774,String.replaceSubrange.RepChar,1,1209,1209,1209,0,1209
775,String.replaceSubrange.RepChar.Small,1,186,186,186,0,186
776,String.replaceSubrange.String,1,54,54,54,0,54
777,String.replaceSubrange.String.Small,1,52,52,52,0,52
778,String.replaceSubrange.Substring,1,122,122,122,0,122
779,String.replaceSubrange.Substring.Small,1,63,63,63,0,63

Total performance tests executed: 8
```

milseman commented:
@keitaito there are explicit pre-specializations for the others: https://github.com/apple/swift/blob/master/stdlib/public/core/StringRangeReplaceableCollection.swift#L197

Benchmarking a non-pre-specialized argument type helps identify and track changes in the fully-generic implementation.

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
I’m running the benchmarks on CI with your current multiplier (5000), to get a relation to your local results.

milseman commented:
@palimondo if @keitaito fixes the workload size as recommended in the report, are we good to go?

-- | --
-- | --
⛔️⏱ | String.replaceSubrange.RepChar execution took at least 17589 μs. Decrease the workload of String.replaceSubrange.RepChar by a factor of 32 (100), to be less than 1000 μs.
⚠️⏱ | String.replaceSubrange.RepChar.Small execution took at least 2589 μs. Decrease the workload of String.replaceSubrange.RepChar.Small by a factor of 4 (10), to be less than 1000 μs.
⚠️⏱ | String.replaceSubrange.Substring execution took at least 1047 μs. Decrease the workload of String.replaceSubrange.Substring by a factor of 2 (10), to be less than 1000 μs.

palimondo commented:
@milseman If you don't foresee an order of magnitude improvement in the performance of `replaceSubrange` for the fast cases (`String`, `Substring`, …) I'd say we should go with the same multiplier for the whole family. It looks like something in 500—1000 range would work fine based on the report above, with `RepChar` jumping above 1000μs, which is still fine so that we can have direct comparison between all the cases.

Otherwise he should go with your recommendation [from before](https://github.com/apple/swift/pull/25310#discussion_r300794907) to:
> … refactor this to a generic `@inline(__always)` implementation function that takes a scale factor (because the String version should be much faster than the Repeat version) and does the loop and `replaceSubrange` calls.

If I understand the point correctly, inlined function will allow the compiler to unroll the loop, which would not be the case if the loop multiplier was passed as parameter to non-inlined generic function? (I don't think we'd need to create per-case functions, as your next sentence said, because the inline closures play the same role and there is nowhere they can get inlined into… I think the ritualistic way we sprinkle `@inline(never)` is obsolete there.)

milseman commented:
Same multiplier sounds good to me if we can make all the benchmarks fit in our time windows

palimondo committed:
[benchmark] String.replaceSubrange: multiplier fix

+[Gardening] minor formatting (80 chars)

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please benchmark

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@swift-ci please smoke test

palimondo commented:
@keitaito Thank you for your patience and your contribution!

keitaito commented:
Woot! Thank you Pavol for updating the multiplier for me, and thank you for merging the PR! I really appreciate both Pavol and Michael that you patiently and carefully gave me reviews and feedback on my PR. Sorry for taking a long time to get it merged, but I'm happy and excited it's finally merged 🙂 I learned very interesting things under the hood of Swift standard library and String APIs through working on it. Thank you so much!

keitaito commented:
I have a question for @milseman about pre-specializations. 

> Benchmarking a non-pre-specialized argument type helps identify and track changes in the fully-generic implementation.

Does this mean that if `@_specialize(where C == Repeated<Character>)` is added to `replaceSubrange(_:with:)`, we could see performance improvement for it when the argument is of type `Repeated<Character>`? Is there any specific reason the pre-specializations annotation is not added unlike `String`, `Substring`, and `Array<Character>` right now?

keitaito commented:
Another question: would you mind explaining what's the criteria for choosing `@inline(never)` vs `@inline(__always)` annotations? According to [Bruno's blog post][1] and [Slava's explanation on Swift mailing list][2], it seems these work like the following:

- If `@inline(never)` is used for the given function, the function won't be inlined. In other words, the function won't be copied to the call site. There could be function call overhead.
- If `@inline(__always)` is used for the given function, the function will be inlined at the call site. No function call overhead, but if the function is called from multiple call sites, there will be code duplications, and the binary size will increase.

Given these, could inlining affect benchmark performance? Looking at benchmark source code files, it seems `@inline(__always)` usage is much less than `@inline(never)`. What kind of function could be annotated with `@inline(__always)`?

[1]: https://swiftrocks.com/the-forbidden-inline-attribute-in-swift.html
[2]: https://lists.swift.org/pipermail/swift-users/Week-of-Mon-20170227/004886.html

palimondo commented:
The ritualistic placement of `@inline(never)` on the tested function [predates the public history](https://github.com/atrick/swift/commit/0356ec8ec373e64b2100adb939f9a1dbfb9cd9bc#diff-abe3f6b9eac1f41e9d42acc8e3e711d7R23) of  Swift Benchmark Suite. I can only guess that it stemmed from some very early internal version of the benchmark where it could have been inlined into the measurement loop? AFAIK it has no practical effect, because the individual benchmark functions are always executed indirectly through a function reference stored in a dictionary – there’s nowhere these functions can be inlined into as far as compiler can see.

For a more targeted use of `@inline(never)` see [`DataBenchmarks`](https://github.com/apple/swift/blob/master/benchmark/single-source/DataBenchmarks.swift#L229) where it is used to explicitly control the inlining into the `runFunction` closures to create 2 different variants of benchmarks to showcase the effect of their (`Data`) inlining. (But frankly I’m not 100% positive I got all that correctly – AFAIK `Data` family of benchmarks is first one to deal explicitly with that aspect.)

For an example of technically correct `@inline(__always)` use, see [`InsertCharacter`](https://github.com/apple/swift/blob/1c0778bb5b17321a2089849ccb167101cbff82c4/benchmark/single-source/InsertCharacter.swift#L32) – where it ensures that the multiple benchmark functions are testing the same thing. But that pattern is not worth emulating, because the same effect can be achieved by leaving out all the annotations and calling the benchmark function with varying parameters from `runFunction` closure.

So I don’t see a very good reason to use `@inline(__always)` in context of benchmark definitions. If you were asking generally, see all occurrences of this annotation in `stdlib`.

palimondo commented:
@keitaito If you’re interested in playing some more with this, I think it should be possible and even beneficial — to prevent further confusion — to remove most of the `@inline(never)` annotations from Swift Benchmark Suite. Another starter task?
